module-system/module-implementations/sov-attester-incentives/src/call.rs

```diff
@@ -0,0 +1,764 @@
+                    working_set,
+                ));
+            }
+
+            // Get the user's old balance.
+            // Transfer the bond amount from the sender to the module's address.
+            // On failure, no state is changed
+            self.reward_sender(context, unbonding_info.amount, working_set)?;
+
+            // Update our internal tracking of the total bonded amount for the sender.
+            self.bonded_attesters.remove(context.sender(), working_set);
+            self.unbonding_attesters
+                .remove(context.sender(), working_set);
+
+            // Emit the unbonding event
+            working_set.add_event("unbonded_challenger", {
+                let amount = unbonding_info.amount;
+                &format!("amount_withdrawn: {:?}", amount)
+            });
+        } else {
+            return Err(self.slash_burn_reward(
+                context.sender(),
+                Role::Attester,
+                SlashingReason::AttesterIsNotUnbonding,
+                working_set,
+            ));
+        }
+        Ok(CallResponse::default())
+    }
+
+    /// The bonding proof is now a proof that an attester was bonded during the last `finality_period` range.
+    /// The proof must refer to a valid state of the rollup. The initial root hash must represent a state between
+    /// the bonding proof one and the current state.
+    fn check_bonding_proof(
+        &self,
+        context: &C,
+        attestation: &Attestation<StorageProof<<C::Storage as Storage>::Proof>>,
+        working_set: &mut WorkingSet<C::Storage>,
+    ) -> Result<(), AttesterIncentiveErrors> {
+        let bonding_root = {
+            // If we cannot get the transition before the current one, it means that we are trying
+            // to get the genesis state root
+            if let Some(transition) = self
+                .chain_state
+                .historical_transitions
+                .get(&(attestation.proof_of_bond.transition_num - 1), working_set)
+            {
+                transition.post_state_root()
+            } else {
+                self.chain_state
+                    .genesis_hash
+                    .get(working_set)
+                    .expect("The genesis hash should be set at genesis")
+            }
+        };
+
+        // This proof checks that the attester was bonded at the given transition num
+        let bond_opt = working_set
+            .backing()
+            .verify_proof(
+                bonding_root,
+                attestation.proof_of_bond.proof.clone(),
+                context.sender(),
+                &self.bonded_attesters,
+            )
+            .map_err(|_err| AttesterIncentiveErrors::InvalidBondingProof)?;
+
+        let bond = bond_opt.unwrap_or_default();
+        let bond: u64 = BorshDeserialize::deserialize(&mut bond.value())
+            .map_err(|_err| AttesterIncentiveErrors::InvalidBondFormat)?;
```

Comment:
nit: This will cause a slightly confusing error message if `bond_opt` is `None`. The returned error will be `InvalidBondFormat` when it should be `UserNotBonded`. We can fix it by using roughly `bond_opt.map(BorshDeserialize).ok_or(Err(AttesterIncentiveErrors::UserNotBonded))??;`